### PR TITLE
streaming csv support

### PR DESCRIPTION
**Pull request by ryantxu**
This takes the existing simple csv parsing and makes it complex :)  Now it finds the series field types and makes them match.  It also adds a way to explicitly set the field properties from CSV.  Using headers like:
```
#name#a,b,c
#type#number,string,boolean
#unit#ms,,s
```
If the special #{field}# header is not present, it uses the first line as the name if the first value is not a number.

This was developed to support #16059, and has been tested with that work in progress

My real goal is to have a solid/robust/supported way to read/write TableData[].  Obviously JSON works great, but it does not work well for large datasets and requires special libraries for streaming.  CSV is super simple and portable.

### CODE REVIEW

File: public/app/plugins/datasource/streaming/query_ctrl.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  }
+
+  // Called from the search window
+  onKeyPress(event: any) {
+    if (event.which === 13) {
+      this.panelCtrl.refresh();
+    }
+  }
+
+  onDataReceived = dataList => {
```
Comment (torkelo) [inline public/app/plugins/datasource/streaming/query_ctrl.ts]: this is the only (I think) remaining puzzle piece for react query editors. Just gave it a quick stab: https://github.com/grafana/grafana/pull/16233

**Review by torkelo — COMMENTED** [review]
cool stuff, do you have a streaming backend you test with. This makes me want to revive the websocket stuff in the backend I worked on 3 years ago (still there)

File: public/app/plugins/datasource/streaming/plugin.json [hunk 2]
```diff
@@ -0,0 +1,19 @@
+{
+  "type": "datasource",
+  "name": "Streaming",
+  "id": "streaming",
+
+  "state": "alpha",
```
[the remark this answers is not in the document]
Comment (ryantxu) [reply public/app/plugins/datasource/streaming/plugin.json]: Ya -- right now it gives you a warning when you create the datasource.

**Comment by ryantxu** [comment]
For now I have merged #16059 with this issue since the reason for the parsing complexity is to support streaming.

If the CSV part looks reasonable, I will move it back to the other branch.

Here is this patch streaming three signals from the included test server:
Screen Recording 2019-03-26 at 11 01 02

File: public/app/plugins/datasource/streaming/query_ctrl.ts (hunk 1 above)
Comment (ryantxu) [reply public/app/plugins/datasource/streaming/query_ctrl.ts]: this particular line is just copy/paste.  The streaming server config options need a lot of love

**Comment by ryantxu** [comment]
@torkelo -- I have made this patch only the CSV parsing stuff, and will continue to work on #16059.  This CSV parsing enforces that the data in a column matches the type.  I think that is a good constraint :)  Previously each value was checked if it could be converted to a number and was if it could be.

**Comment by ryantxu** [comment]
eventually this should replace most of https://github.com/grafana/grafana/blob/master/public/app/core/utils/file_export.ts#L57

the existing CSV format is super agressive with quoting and escaping -- this version checks that the value needs escaped before adding quotes and escaping.

**Comment by torkelo** [comment]
csv support in the testdata data source would be nice. But a bit tricky without supporting SeriesData in the backend metrics api / models. 

The easiest way would be to parse data in frontend and transform to TableData and save to query model so backend can return it. But will lose units & type as that is not supported in the backend table model.

**Comment by ryantxu** [comment]
I was thinking a frontend only datasource that just data in the panel/datasource json and passes it back directly.

Super useful for the times you just need to hardcode some values.

It could be in testdata -- but I think it makes sense as its own thing.

**Comment by torkelo** [comment]
Yea, agree can be it’s own

**Pull request merged**